### PR TITLE
Add elastic-agent-service container to packaging

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Adds a new docker variant of `elastic-agent-service` that includes the python connectors.

Currently this PR includes the following PR until that PR is merged - https://github.com/elastic/elastic-agent/pull/5338

## Why is it important?

Used by the Elastic Agent service.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~ (not really user facing at this level)
- ~~[ ] I have added an integration test or an E2E test~~

## Disruptive User Impact

None.

## How to test this PR locally

```
$ PLATFORMS="linux/arm64" PACKAGES="docker" SNAPSHOT="true" EXTERNAL="true" DOCKER_VARIANTS="service" mage package
$ docker run -it --rm docker.elastic.co/beats-ci/elastic-agent-service:8.16.0-SNAPSHOT
```

### CODE REVIEW

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (seanstory) [inline dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:182]: Would be safer to unzip and  `make install`, as that'll also give it a python virtual environment, and keep it resileint to any other system python deps that might end up in this image one day.

**Review by seanstory — COMMENTED** [review]
Few nits but this is super exciting. Thank you so much for your help, it would have taken us ages to figure out how to stitch into this system. 🙏

File: dev-tools/packaging/packages.yml
Comment (blakerouse) [reply dev-tools/packaging/packages.yml:302]: Yeah it doesn't matter. Just need to know what the contents should be.

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (blakerouse) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:181]: I can lower the version, once we know whats in the execution script. By the time its ready you might support 3.12 :-)

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (blakerouse) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:182]: I can do that, seems a little un-needed being this is the only python used in the image. The simple `pip install` without requiring unzip to be installed in the image seems better to me.

**Comment by blakerouse** [comment]
Just note I am blocked on making any more progress on this PR, under we know what the contents of the bash script should be and I believe we are waiting for the python version of the elastic-agent-client to be included in the DRA or possibly in a pypi dist (some way of getting it).

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (seanstory) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:181]: I don't expect 3.12 support to make it that fast. :) We were on 3.10 for about 2 years before we _just_ added 3.11 a few weeks ago. Urgency is not there.

**Comment by seanstory** [comment]
> I am blocked on making any more progress on this PR, under we know what the contents of the bash script should be

Can we not merge it as is, and follow up later with edits to make the binary more "real"? Is there some sort of testing bar that happens for each included binary that we need to to be able to meet?

> I believe we are waiting for the python version of the elastic-agent-client to be included in the DRA

@artem-shelkovnikov is currently working on that, but is needing to do [some repo restructuring](https://github.com/elastic/search-team/issues/8165) so that the private-repo py-elastic-agent-client lib can be installed in this docker image, but won't prevent community members from cloning and building self-managed connectors from source.

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl
Comment (seanstory) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:182]: it's best practice (AFAIU) to not use system python if you can help it. Also by using a `make` target, it allows us to hook into other setup bits. For example, our `make isntall` insures that `bin/pip install --upgrade pip` and `bin/pip install --upgrade setuptools` are both run, and also ensure that the right `requirements/$ARCH.txt` is selected, depending upon the OS that it's being run on.

**Comment by artem-shelkovnikov** [comment]
👋 We should be unblocked on making progress on this PR.

We've recently merged https://github.com/elastic/connectors/pull/2787 that has a component ready to be ran.

To run the component you'd need:
1. Pull the repo
2. Run `make install-agent` in the root of the repo
3. Use python from the virtual environment to run the component: `.venv/bin/python connectors/agent/cli.py`

I tested this code locally by having an agent downloaded and adding component and component spec into the components with following code (would need to be adapted to run in the docker image):

```
artemshelkovnikov-m2@Work-M2 components % cat connectors-agent
#!/bin/bash
FOLDER=/Users/artemshelkovnikov-m2/git_tree/connectors-py/.venv/bin

$FOLDER/elastic-agent-connectors $FOLDER/elastic-agent-connectors

artemshelkovnikov-m2@Work-M2 components % cat connectors-agent.spec.yml
version: 2
inputs:
  - name: connectors-py
    description: "Connectors Py component input"
    platforms: &platforms
      - linux/amd64
      - linux/arm64
      - darwin/amd64
      - darwin/arm64
      - windows/amd64
      - container/amd64
      - container/arm64
    outputs: &outputs
      - elasticsearch
    shippers: &shippers
      - shipper
    command: &command
      restart_monitoring_period: 5s
      maximum_restarts_per_period: 1
      timeouts:
        restart: 1s
      args: []
```

**Comment by seanstory** [comment]
Woo!! Thanks, Artem!
@blakerouse , we were also talking in a call today with @cmacknz , who suggested that we should make sure that this PR also includes some basic smoke tests that validate that Agent can start the connector binary, and do some basic communication. This will be really valuable for us since we're using a new python agent client, so anything that can catch basic problems earlier in the build/test pipeline is good. :)

**Comment by blakerouse** [comment]
@seanstory I can add an integration test using this container, once I have it building and working.

**Comment by artem-shelkovnikov** [comment]
@blakerouse the problem comes from the fact that connectors dependency is a private Elastic repository.

We had to work around this with this change: https://github.com/elastic/ci/pull/3327

Additionally, we're gonna be opening the private repo ASAP.

I don't think it would work even if we provided a `zip` or any other artifact, as the problem is with access to the dependency?

**Comment by blakerouse** [comment]
@artem-shelkovnikov Opening it to be public would make this much easier and allow developers to build this locally with our simple `mage package` target in the Elastic Agent repository. ETA on public access?

**Comment by artem-shelkovnikov** [comment]
ETA unknown - I've submitted a request to #open-source here: https://github.com/elastic/open-source/issues/443

Open source team will look at it today and best case will allow to open it up, I'll follow up here after I get a response

**Comment by seanstory** [comment]
We[ just got the approval](https://github.com/elastic/open-source/issues/443#issuecomment-2338513554), and https://github.com/elastic/python-elastic-agent-client now has public/open visibility 🎉 
@blakerouse can you try again?

**Comment by seanstory** [comment]
Fixed by https://github.com/elastic/connectors/pull/2811. Sorry about that. New DRAs should be available momentarily. 🤞

**Comment by blakerouse** [comment]
@seanstory

I don't think the DRA has that change or even the change that was merged last week, or something is wrong

The contents of the DRA only show the following

```
bash-5.2$ ls
VERSION                __pycache__            config.py              content_extraction.py  filtering              logger.py              protocol               services               sync_job_runner.py
__init__.py            access_control.py      connectors_cli.py      es                     kibana.py              preflight_check.py     service_cli.py         source.py              utils.py
```

which doesn't have the `agent` or `cli` directory at all

**Comment by seanstory** [comment]
For those watching, we solved the missing dir issue with a three-part combo of:
- https://github.com/elastic/connectors/pull/2811
- https://github.com/elastic/connectors/pull/2812
- https://github.com/elastic/connectors/pull/2814
and now we're discussing further hiccups in [this slack thread](https://elastic.slack.com/archives/C07C6P8AQ8H/p1725988035363169).

**Comment by blakerouse** [comment]
@seanstory @artem-shelkovnikov Seems that the 8.16 DRA finally has the changes needed to get the connector running. My latest push has it running in the container.

I am working on creating an integration test for running the connector, could you provide a configuration that I should use with Elastic Agent that will run the connector and have it connect to elasticsearch and ingest data. I can then write a validation to ensure that:

1. Connector runs under agent
2. It is reporting healthy
3. That it was able to take the output configuration
4. That it wrote data to the output

**Comment by blakerouse** [comment]
Hit an issue writing integration tests for this PR. It was not possible to define a custom docker variant for the integration tests for Kubernetes.

https://github.com/elastic/elastic-agent/pull/5547 fixes this issue and needs to be merged before I can get this PR cleaned up to add an integration test.

**Comment by seanstory** [comment]
looks like https://github.com/elastic/elastic-agent/pull/5547 just merged!

@blakerouse I know you've hit a lot of roadblocks on this PR, so this is not a judgment laden question, I'm just looking to keep my estimated timelines realistic. 🙏 
How much do you feel is left on this? Originally, we'd sequenced our milestones with the assumption that Agentless would be able to switch to the new image by Oct 1. So if there's still more than 1 week of work left here, please let me know so I can start shuffling dates and adjusting downstream expectations.

**Comment by blakerouse** [comment]
@seanstory I finally have a passing integration test for this new container. Its all working now. Just need a review and it can be merged.

File: testing/integration/kubernetes_agent_standalone_test.go
Comment (pkoutsovasilis) [inline testing/integration/kubernetes_agent_standalone_test.go:687]: oh nice! this is helpful!

File: testing/integration/kubernetes_agent_standalone_test.go
Comment (pkoutsovasilis) [inline testing/integration/kubernetes_agent_standalone_test.go:715]: here we want to keep logging the stdout and stderr when agent didn't report healthy right?!

**Comment by pkoutsovasilis** [comment]
hey @blakerouse 👋 I left only a minor comment that I think you need to address. However, may I join forces with you here and push a Helm-chart based integration test? I was looking for a reason to add a custom integration/input test for the agent Helm chart and this one seems like ideal. Or maybe it can be a follow up pull request, no prob with that either 🙂

File: testing/integration/kubernetes_agent_standalone_test.go
Comment (blakerouse) [reply testing/integration/kubernetes_agent_standalone_test.go]: good catch, I will add that back

**Comment by blakerouse** [comment]
@pkoutsovasilis Fixed

**Review by pkoutsovasilis — APPROVED** [review]
LGTM

File: magefile.go
Comment (pchila) [inline magefile.go:1164]: Is there an alternative to adding the PythonWheel binary specs to the `ExpectedBinaries` just to remove them from the flattened directory? The connectors are added to `ExpectedBinaries` just to download the archive ?
If the handling of connectors is different from the other components' binaries we package with agent maybe we can separate the handling with a `collectAdditionalDependencies` target  after flattening or something similar ? 
My understanding is also that these dependencies are not needed to create an agent archive but get installed when we run `docker build`, maybe we can move the processing at that stage ?

**Comment by blakerouse** [comment]
buildkite test this

**Comment by jlind23** [comment]
buildkite test it

**Comment by blakerouse** [comment]
buildkite test this

**Pull request merged**
